### PR TITLE
SNOW-1805836: Implement the dataframe interchange protocol.

### PR DESCRIPTION
**Pull request by sfc-gh-mvashishtha**
Fixes SNOW-1805836

### CODE REVIEW

File: tests/integ/modin/test_scikit.py [hunk 1]
```diff
@@ -7,39 +7,27 @@
...
     # thrown.
-    with SqlCounter(query_count=0):
-        try:
-            MaxAbsScaler().fit_transform(X)
-            raise AssertionError()
-        except NotImplementedError:
-            # scikit-learn 1.5.2 will throw an error for the DF
-            # interchange protocol.
-            pass
+    MaxAbsScaler().fit_transform(X)
```
[the remark this answers is not in the document]
Comment (sfc-gh-mvashishtha) [reply tests/integ/modin/test_scikit.py:17]: The new query count of 5 is correct.

I've removed some outdated comments about the test cases here.

File: src/snowflake/snowpark/modin/plugin/extensions/dataframe_overrides.py [hunk 2]
```diff
@@ -754,15 +754,6 @@ def __dataframe__(self, nan_as_null: bool = False, allow_copy: bool = True):
     ProtocolDataframe
```
[the remark this answers is not in the document]
Comment (sfc-gh-mvashishtha) [reply src/snowflake/snowpark/modin/plugin/extensions/dataframe_overrides.py]: Good point. Also, the docstring here has no effect, so I made the change in `docstrings/dataframe.py`.

File: tests/integ/modin/frame/test_interchange_protocol.py [hunk 3]
```diff
@@ -0,0 +1,628 @@
...
+            column = interchange_dataframe.get_column(0)
+            chunks = list(column.get_chunks())
+            assert len(chunks) == 1
+            buffer, _ = column.get_buffers()["data"]
+            return buffer.ptr
+
+        assert get_buffer_memory_pointer(interchange1) != get_buffer_memory_pointer(
+            interchange2
+        )
+
```
Comment (sfc-gh-lmukhopadhyay) [inline tests/integ/modin/frame/test_interchange_protocol.py:391]: Thanks for adding these detailed integ tests! Have you already checked if the relevant native pandas interchange integration tests cases have been included here, such as [test_non_str_names_w_duplicates](https://github.com/pandas-dev/pandas/blob/652682993cfe130c733944f494e82c367bcfac5b/pandas/tests/interchange/test_impl.py#L426C5-L426C36)?

File: CHANGELOG.md [hunk 4]
```diff
@@ -64,6 +64,7 @@
 - Added support for `DataFrame.align` and `Series.align` for `axis=1` and `axis=None`.
 - Added support fot `pd.json_normalize`.
 - Added support for `GroupBy.pct_change` with `axis=0`, `freq=None`, and `limit=None`.
+- Added support for `DataFrameGroupBy.__iter__` and `SeriesGroupBy.__iter__`.
```
Comment (sfc-gh-rdurrani) [inline CHANGELOG.md:67]: Merge artifact?

File: src/snowflake/snowpark/modin/plugin/docstrings/dataframe.py [hunk 5]
```diff
@@ -5099,7 +5099,7 @@ def __delitem__():
 
     def __dataframe__():
         """
-        Get a Modin DataFrame that implements the dataframe exchange protocol.
+        Get an object that implements the dataframe interchange protocol for this dataframe.
```
Comment (sfc-gh-rdurrani) [inline src/snowflake/snowpark/modin/plugin/docstrings/dataframe.py:5102]: Maybe add a note that the data will be pulled into local memory? Doesn't hurt to be explicit.

**Review by sfc-gh-rdurrani — COMMENTED** [review]
left two minor nits, overall lgtm

File: tests/integ/modin/test_scikit.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (sfc-gh-mvashishtha) [reply tests/integ/modin/test_scikit.py:17]: Yep.

File: CHANGELOG.md (hunk 4 above)
Comment (sfc-gh-mvashishtha) [reply CHANGELOG.md:67]: For me, this line doesn't show up in the diff against main.

File: CHANGELOG.md [hunk 6]
```diff
@@ -65,6 +65,9 @@
 - Added support fot `pd.json_normalize`.
 - Added support for `GroupBy.pct_change` with `axis=0`, `freq=None`, and `limit=None`.
 - Added support for `DataFrameGroupBy.__iter__` and `SeriesGroupBy.__iter__`.
+- Added partial support for the dataframe interchange protocol method
+  `DataFrame.__dataframe__()`. Snowpark pandas does not yet support columns of
+  `Timdelta` type or columns containing list objects.
```
[the remark this answers is not in the document]
Comment (sfc-gh-mvashishtha) [reply CHANGELOG.md:70]: I think it's not necessary to do so, but it is helpful.

File: src/snowflake/snowpark/modin/plugin/docstrings/dataframe.py (hunk 5 above)
Comment (sfc-gh-mvashishtha) [reply src/snowflake/snowpark/modin/plugin/docstrings/dataframe.py:5102]: Done.

**Review by sfc-gh-rdurrani — APPROVED** [review]
LGTM!

File: tests/integ/modin/test_scikit.py [hunk 7]
```diff
@@ -7,39 +7,19 @@
...
 
 
-# SNOW-1518382 Support PCA  on earlier versions of
-# scikit-learn by supporting numpy.may_share_memory
-def test_scikit_pca():
+@sql_count_checker(query_count=3)
+def test_pca():
```
Comment (sfc-gh-helmeleegy) [inline tests/integ/modin/test_scikit.py:21]: Aren't there other 3rd party libraries that we know can now benefit from implementing the interchange protocol? If so, we should consider adding integration tests for them as well. Doesn't need to be in this PR though.

**Review by sfc-gh-helmeleegy — APPROVED** [review]
Looks great. Thanks, Mahesh.

File: CHANGELOG.md [hunk 8]
```diff
@@ -65,6 +65,9 @@
 - Added support fot `pd.json_normalize`.
 - Added support for `GroupBy.pct_change` with `axis=0`, `freq=None`, and `limit=None`.
 - Added support for `DataFrameGroupBy.__iter__` and `SeriesGroupBy.__iter__`.
+- Added partial support for the dataframe interchange protocol method
+  `DataFrame.__dataframe__()`. Snowpark pandas does not yet support columns of
```
Comment (sfc-gh-yzou) [inline CHANGELOG.md:69]: I thought we support timedelta type? I see the following branch in to_pandas 
```
 if (
                snowpark_pandas_type is not None
                and snowpark_pandas_type == TimedeltaType()
            ):
```

Comment (sfc-gh-yzou) [reply CHANGELOG.md:69]: is the limitation same as the Snowpark pandas to_pandas limitaiton? if that is the case, maybe just refer to the limitation of to_pandas here

File: docs/source/modin/supported/general_supported.rst [hunk 9]
```diff
@@ -25,6 +25,9 @@ Data manipulations
 +-----------------------------+---------------------------------+----------------------------------+----------------------------------------------------+
 | ``cut``                     | P                               | ``retbins``, ``labels``          | ``N`` if ``retbins=True``or ``labels!=False``      |
 +-----------------------------+---------------------------------+----------------------------------+----------------------------------------------------+
+| ``__dataframe__``           | P                               |                                  | ``N`` for columns of type ``Timedelta`` and columns|
+|                             |                                 |                                  | containing list objects                            |
```
Comment (sfc-gh-yzou) [inline docs/source/modin/supported/general_supported.rst:29]: what does N mean here, in general N will throw a NotImplementedError, is that the same here?

File: src/snowflake/snowpark/modin/plugin/docstrings/dataframe.py [hunk 10]
```diff
@@ -5099,10 +5099,13 @@ def __delitem__():
...
 
         See more about the protocol in https://data-apis.org/dataframe-protocol/latest/index.html.
 
+        Like `DataFrame.to_pandas`, this method collects all this dataframe's
+        data.
```
Comment (sfc-gh-yzou) [inline src/snowflake/snowpark/modin/plugin/docstrings/dataframe.py:5107]: -> this methods triggers an evaluation and pulls data to the local machine

File: tests/integ/modin/test_scikit.py (hunk 7 above)
Comment (sfc-gh-mvashishtha) [reply tests/integ/modin/test_scikit.py:21]: There are more such third party libraries, and we also want to extend this scikit-learn test suite. We have follow-up tasks to add more tests.

File: tests/integ/modin/frame/test_interchange_protocol.py (hunk 3 above)
Comment (sfc-gh-mvashishtha) [reply tests/integ/modin/frame/test_interchange_protocol.py:391]: Thanks for bringing that test file up. I looked through it and I think I added the relevant tests.

**Review by sfc-gh-lmukhopadhyay — APPROVED** [review]
LGTM, thanks Mahesh!

File: CHANGELOG.md [hunk 11]
```diff
@@ -66,6 +66,9 @@
 - Added support for `GroupBy.pct_change` with `axis=0`, `freq=None`, and `limit=None`.
 - Added support for `DataFrameGroupBy.__iter__` and `SeriesGroupBy.__iter__`.
 - Added support for `np.sqrt`, `np.trunc`, `np.floor`, numpy trig functions, `np.exp`, `np.abs`, `np.positive` and `np.negative`.
+- Added partial support for the dataframe interchange protocol method
+  `DataFrame.__dataframe__()`. Snowpark pandas does not yet support columns of
+  `Timdelta` type or columns containing list objects.
```
Comment (sfc-gh-jkew) [inline CHANGELOG.md:71]: "Timdelta" => "Timedelta"

Comment (sfc-gh-mvashishtha) [reply CHANGELOG.md:71]:
```suggestion
  `Timedelta` type or columns containing list objects.
```

File: CHANGELOG.md (hunk 8 above)
Comment (sfc-gh-yzou) [reply CHANGELOG.md:70]: sorry, I am confused about this part, i see the implementation is the following 
```
return self.to_pandas().__dataframe__(
            nan_as_null=nan_as_null, allow_copy=allow_copy
        )
```

do you mean the pandasDF.__dataframe__ doesn't support timedelta, what happens when there is a timedelta column after to_pandas(), does it throws an error?
In order to add full support, do you need to go add support in pandas?

Comment (sfc-gh-yzou) [reply CHANGELOG.md:70]: if that is the case, i don't think we have mention this in the release log, you did a support of DataFrame.__dataframe__(). 
However, mention those limitaitons in the doc under docstrings/dataframe.py, and be clear that those are limitations from https://pandas.pydata.org/docs/reference/api/pandas.DataFrame.__dataframe__.html

Comment (sfc-gh-mvashishtha) [reply CHANGELOG.md]: I disagree, but not strongly enough to engage in a debate. Done.

**Review by sfc-gh-yzou — APPROVED** [review]

**Pull request merged**